### ISSUE TITLE
java.lang.ClassNotFoundException: com.google.errorprone.BaseErrorProneJavaCompiler

### ISSUE THREAD

**Issue by hakanai** [tbroyer/gradle-errorprone-plugin#15]
We are successfully using errorprone 0.0.16 in a build, and now trying to update to 0.6, which fails with a `ClassNotFoundException` when trying to compile the main classes.

Is there something missing from my configuration? The one thing I wasn't sure about from reading the docs was the version I was supposed to be putting in for errorproneJava.

This is the configuration we ended up with:

```
buildscript {
    repositories {
        // ...
    }

    dependencies {
        classpath 'net.ltgt.gradle:gradle-errorprone-plugin:0.6'
    }
}

apply plugin: 'java'
apply plugin: net.ltgt.gradle.errorprone.ErrorPronePlugin

dependencies {
    errorprone 'com.google.errorprone:error_prone_core:2.3.2'
    errorproneJavac 'com.google.errorprone:javac:9+181-r4173-1'
}

import net.ltgt.gradle.errorprone.CheckSeverity

tasks.withType(JavaCompile) {
    options.errorprone {
        disableWarningsInGeneratedCode = true

        check 'BadComparable', CheckSeverity.ERROR
        // ...
    }
}
```

When running the build, I get the following output:

```
$ ./gradlew check
Parallel execution is an incubating feature.

> Task :compileJava FAILED
warning: [options] bootstrap class path not set in conjunction with -source 1.8
An exception has occurred in the compiler (9-internal). Please file a bug against the Java compiler via the Java bug reporting page (http://bugreport.java.com) after checking the Bug Database (http://bugs.java.com) for duplicates. Include your program and the following diagnostic in your report. Thank you.
java.lang.NoClassDefFoundError: com/google/errorprone/BaseErrorProneJavaCompiler
        at com.google.errorprone.ErrorProneJavacPlugin.init(ErrorProneJavacPlugin.java:39)
        at com.sun.tools.javac.api.BasicJavacTask.initPlugins(BasicJavacTask.java:214)
        at com.sun.tools.javac.api.JavacTaskImpl.prepareCompiler(JavacTaskImpl.java:192)
        at com.sun.tools.javac.api.JavacTaskImpl.lambda$doCall$0(JavacTaskImpl.java:97)
        at com.sun.tools.javac.api.JavacTaskImpl.handleExceptions(JavacTaskImpl.java:142)
        at com.sun.tools.javac.api.JavacTaskImpl.doCall(JavacTaskImpl.java:96)
        at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:90)
        at org.gradle.api.internal.tasks.compile.AnnotationProcessingCompileTask.call(AnnotationProcessingCompileTask.java:89)
        at org.gradle.api.internal.tasks.compile.ResourceCleaningCompilationTask.call(ResourceCleaningCompilationTask.java:57)
        at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:50)
        at org.gradle.api.internal.tasks.compile.JdkJavaCompiler.execute(JdkJavaCompiler.java:36)
        at org.gradle.api.internal.tasks.compile.daemon.AbstractDaemonCompiler$CompilerCallable.call(AbstractDaemonCompiler.java:88)
        at org.gradle.api.internal.tasks.compile.daemon.AbstractDaemonCompiler$CompilerCallable.call(AbstractDaemonCompiler.java:76)
        at org.gradle.workers.internal.DefaultWorkerServer.execute(DefaultWorkerServer.java:42)
        at org.gradle.workers.internal.WorkerDaemonServer.execute(WorkerDaemonServer.java:46)
        at org.gradle.workers.internal.WorkerDaemonServer.execute(WorkerDaemonServer.java:30)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.gradle.process.internal.worker.request.WorkerAction.run(WorkerAction.java:101)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:35)
        at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
        at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:155)
        at org.gradle.internal.remote.internal.hub.MessageHubBackedObjectConnection$DispatchWrapper.dispatch(MessageHubBackedObjectConnection.java:137)
        at org.gradle.internal.remote.internal.hub.MessageHub$Handler.run(MessageHub.java:404)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:63)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:46)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:55)
        at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.ClassNotFoundException: com.google.errorprone.BaseErrorProneJavaCompiler
        at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        ... 36 more

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':compileJava'.
> Compilation failed; see the compiler error output for details.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1s
1 actionable task: 1 executed
```

More diagnostics:

```
$ ./gradlew dependencies --configuration errorprone
...
errorprone
\--- com.google.errorprone:error_prone_core:2.3.2
     +--- com.google.errorprone:error_prone_annotation:2.3.2
     |    \--- com.google.guava:guava:23.5-jre
     |         +--- com.google.code.findbugs:jsr305:1.3.9 -> 3.0.0
     |         +--- org.checkerframework:checker-qual:2.0.0 -> 2.5.3
     |         +--- com.google.errorprone:error_prone_annotations:2.0.18 -> 2.3.2
     |         +--- com.google.j2objc:j2objc-annotations:1.1
     |         \--- org.codehaus.mojo:animal-sniffer-annotations:1.14
     +--- com.google.errorprone:error_prone_type_annotations:2.3.2
     +--- com.google.errorprone:error_prone_check_api:2.3.2
     |    +--- com.google.errorprone:error_prone_annotation:2.3.2 (*)
     |    +--- com.google.code.findbugs:jsr305:3.0.0
     |    +--- org.checkerframework:dataflow:2.5.3
     |    |    +--- org.checkerframework:checker-qual:2.5.3
     |    |    \--- org.checkerframework:javacutil:2.5.3
     |    |         \--- org.checkerframework:checker-qual:2.5.3
     |    +--- com.googlecode.java-diff-utils:diffutils:1.3.0
     |    +--- com.google.errorprone:error_prone_annotations:2.3.2
     |    \--- com.github.kevinstern:software-and-algorithms:1.0
     +--- com.github.stephenc.jcip:jcip-annotations:1.0-1
     +--- org.pcollections:pcollections:2.1.2
     +--- com.google.guava:guava:23.5-jre (*)
     +--- com.google.auto:auto-common:0.10
     |    \--- com.google.guava:guava:23.5-jre (*)
     +--- com.google.code.findbugs:jFormatString:3.0.0
     +--- com.google.code.findbugs:jsr305:3.0.0
     +--- org.checkerframework:dataflow:2.5.3 (*)
     +--- com.google.errorprone:error_prone_annotations:2.3.2
     \--- com.google.protobuf:protobuf-java:3.4.0
```

```
$ ./gradlew dependencies --configuration errorproneJavac
...

errorproneJavac
\--- com.google.errorprone:javac:9+181-r4173-1
```

**Comment by tbroyer** [comment]
`ErrorProneJavacPlugin` is in `error_prone_core`, whereas `BaseErrorProneJavaCompiler` is in `error_prone_check_api`. Can you check your `./gradlew dependencies --configuration annotationProcessor` to see if it correctly includes `error_prone_check_api`?

Do you have a [MCVE](https://stackoverflow.com/help/mcve)/[SSCCE](http://sscce.org/) so I could try reproducing?

**Comment by hakanai** [comment]
The hint that it's the annotationProcessor configuration was enough to figure out what was going on.

For compile configurations, we were blocking transitive dependencies, which is reasonable. We were treating annotationProcessor like a compile configuration, which breaks if you do this since it's actually running stuff. I have now removed annotationProcessor (and testAnnotationProcessor) from having that done to it, and things seem fine.

I did have to jump through some hoops to force some dependencies up to the versions errorprone had, because they were clashing with each other.

```
> A conflict was found between the following modules:
    - com.google.code.findbugs:jsr305:3.0.0
    - com.google.code.findbugs:jsr305:1.3.9
  A conflict was found between the following modules:
    - com.google.errorprone:error_prone_annotations:2.3.2
    - com.google.errorprone:error_prone_annotations:2.0.18
  A conflict was found between the following modules:
    - org.checkerframework:checker-qual:2.0.0
    - org.checkerframework:checker-qual:2.5.3
```

We don't depend on any of these directly ourselves, but I guess errorprone and something else it depends on are clashing on these. Oh well, I'm just forcing them to the latest and hoping for the best.
